### PR TITLE
replication: warn if insufficient workers

### PR DESCRIPTION
**Pull request by poornas**
This should give an early warning if configured replication workers
is insufficient to meet application workload.

## Motivation and Context
Better maintainability

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

File: cmd/bucket-replication.go [hunk 1]
```diff
@@ -1425,6 +1425,7 @@ func (p *ReplicationPool) queueReplicaFailedTask(ri ReplicateObjectInfo) {
 		})
 	case p.mrfReplicaCh <- ri:
 	default:
+		logger.LogOnceIf(GlobalContext, fmt.Errorf("WARNING: Replication failed workers could not keep up with healing failures - consider increasing number of replication failed workers with `mc admin config set api replication_failed_workers=`"), replicationSubsystem)
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]: perhaps even suggest a value based on the current value? @poornas

File: cmd/bucket-replication.go [hunk 2]
```diff
@@ -1450,6 +1451,7 @@ func (p *ReplicationPool) queueReplicaTask(ri ReplicateObjectInfo) {
 		})
 	case ch <- ri:
 	default:
+		logger.LogOnceIf(GlobalContext, fmt.Errorf("WARNING: Replication workers could not keep up with incoming traffic - consider increasing number of replication workers with `mc admin config set api replication_workers=`"), replicationSubsystem)
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]: perhaps even suggest a value based on the current value? @poornas

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#f8da1a28-9ede-4636-866a-bc246cf24e43) |
<a name="f8da1a28-9ede-4636-866a-bc246cf24e43"></a>
### 13899-ff53e0b/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s8:30366
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 55 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 23 seconds
(8/15) Running minio-dotnet tests ... done in 50 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 6 seconds
(10/15) Running minio-java tests ... done in 29 seconds
(11/15) Running minio-js tests ... FAILED in 55 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook for \"removeObject(bucketName, objectName, removeOpts)",
  "args": "bucketName:minio-js-test-ver-presign47016f90-e1d3-4aaf-a9dc-82a2649c1bfe ",
  "duration": 5,
  "alert": "datafile-1-b ",
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty. You must delete all versions in the bucket. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:89:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:164:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:670:10) at endReadableNT (internal/streams/readable.js:1334:12) at processTicksAndRejections (internal/process/task_queues.js:82:21)"
}
(11/15) Running minio-py tests ... done in 1 minutes and 16 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-replication.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-replication.go]: updated the PR.

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**